[mlir][bufferization] Empty tensor elimination for materialize_in_destination

matthias-springer opened the pull request:
This revision adds support for empty tensor elimination to "bufferization.materialize_in_destination" by implementing the `SubsetOpInterface`.

Furthermore, the One-Shot Bufferize conflict detection is improved for "bufferization.materialize_in_destination".

srcarroll commented on mlir/include/mlir/Dialect/Bufferization/IR/InferDestinationOpInterface.td:
```diff
@@ -0,0 +1,103 @@
+        /*desc=*/[{
+          Return the SSA values that are needed to build the destination tensor.
+          This interface method will be called only on tensor operands for which
+          `isAnchor` returns "true"
```
maybe it's just me, but it's unclear what "needed to build" means. Are these all the SSA values that would be required to build whatever op is being built in `getOrBuildDestination`?  I must be thinking about that incorrectly because I see no dependence between the two in the `InsertSliceLikeOpInferDestinationInterface` changes.

matthias-springer replied:
I expanded the documentation and added an example.

srcarroll commented on mlir/include/mlir/Dialect/Bufferization/IR/InferDestinationOpInterface.td:
```diff
@@ -0,0 +1,103 @@
+      InterfaceMethod<
+        /*desc=*/[{
+          Return "true" if the given operand is an "input" tensor on which
+          empty tensor elimination can be applied.
```
to clarify for my own sake, this would be the DPS input?

matthias-springer replied:
The interface is now called `SubsetOpInterface` and this function has turned into `getSource`.

srcarroll commented on mlir/include/mlir/Dialect/Bufferization/IR/InferDestinationOpInterface.td:
```diff
@@ -0,0 +1,103 @@
+  let description = [{
+    This interface can be implemented by ops that read data from an "input"
+    tensor and store the result into an "output"/"init" tensor (i.e., the
+    "destination" tensor). It drives the empty tensor elimination pass.
```
would you say the tensor elimination pass would be the only use for this interface? Or do you think this will have other uses later?

matthias-springer replied:
The original version of this interface was used only for empty tensor elimination. The new SubsetOpInterface is also used by One-Shot Analysis. In the future, it could also be used by the tiling infrastructure (and then move to `mlir/Interfaces`).

matthias-springer commented on mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-analysis.mlir:
```diff
@@ -158,3 +158,31 @@ func.func @bbarg_of_unknown_op_2(%f: f32) {
+  %buffer = tensor.empty(%sz) : tensor<?xf32>
+  // CHECK: bufferization.materialize_in_destination
+  // CHECK-SAME: {__inplace_operands_attr__ = ["true", "true"]}
+  %r = bufferization.materialize_in_destination %buffer in %buffer : tensor<?xf32>
```
The point of the test is that no OpOperand of this op should bufferize out-of-place. In that case, we would end up with 2 memcpy, when in fact there should be no memcpys (because source and destination are the same).

This is essentially a test for the `BufferizableOpInterface` implementation of `MaterializeInDestinationOp`. (Ops that read and write two tensors via separate are usually seen as a RaW conflict and bufferize out-of-place.)

matthias-springer commented:
I rewrote large parts of this stack of commits. `InferDestinationOpInterface` is now called `SubsetOpInterface` and has better documentation.

srcarroll commented on mlir/include/mlir/Dialect/Bufferization/IR/SubsetOpInterface.td:
```diff
@@ -0,0 +1,159 @@
+
+    Ops that deal with tensor subsets come in two flavours:
+    - Insertion flavor: Ops that insert a source tensor into a destination
+      tensor at the specified subset. Such ops usuallt return a new destination
```
typo: usuallt -> usually

srcarroll commented on mlir/include/mlir/Dialect/Bufferization/IR/SubsetOpInterface.td:
```diff
@@ -0,0 +1,159 @@
+        /*methodName=*/"isEquivalentSubset",
+        /*args=*/(ins
+            "::mlir::Value":$candidate,
+            "::llvm::function_ref<bool(Value, Value)>":$equivalenceFn)
```
can you explain what `equivalenceFn` is in the description.  It's unclear to me whether the main logic for determining equivalent subset should be in `equivalenceFn` or `isEquivalentSubset`.   Maybe I'm just missing something, but seems awkward to have both

matthias-springer replied:
I expanded the description in #65619. `equivalenceFn` returns "true" if two values are equivalent. When used in One-Shot Bufferize, I use `AnalysisState::areEquivalentBufferizedValues` as `equivalenceFn`. I didn't want to make this interface dependent on other bufferization stuff (such as `AnalysisState`), so that it can be moved to `mlir/Interfaces` in the future.

srcarroll replied:
there's context I'm missing since I'm new to some of this stuff, so I won't harp on this much longer.  But I guess I still don't see the need for the hook.  Why can't `isEquivalentSubset` be implemented without it? For example, the default could be
```
bool isEquivalentSubset(Value candidate) {
  return getSubset() == candidate;
```
(sorry if this is a naive question)

matthias-springer replied:
This would work for `bufferization.materialize_in_destination` because the subset is simply the destination tensor. But `getSubset` may build new IR. E.g., for `tensor.insert_slice`, it will build a new `tensor.extract_slice` op and return that one. I'm going to rename `getSubset` to `buildSubset`.

srcarroll commented on mlir/include/mlir/Dialect/Bufferization/IR/SubsetOpInterface.td:
```diff
@@ -0,0 +1,159 @@
+    bool isSameSubset(OpResult candidate) {
+      return cast<::mlir::bufferization::SubsetOpInterface>(getOperation())
+          .isEquivalentSubset(candidate,
+                              [](Value v1, Value v2) { return v1 == v2; });
```
It seems a little clunky to solely use `isEquivalentSubset` for determining if "same".  I understand that two subsets are "same" only if they are equivalent, so it makes sense to use `isEquivalentSubset` as part of the logic for this function.  But it seems like you're trying to shove the extra logic for "same" into the `equivalencFn` hook.  Although I may be getting this wrong as I'm not entirely sure of the purpose of the hook (see comment above).

matthias-springer replied:
I wanted to make `equivalenceFn` optional with `[](Value v1, Value v2) { return v1 == v2; }` as as default value for the parameter. But interface methods cannot have optional parameters, so I had to add another function. (Note this is not an InterfaceMethod and cannot be implemented.)

`v1 == v2` is the simplest kind of equivalence function. `isSameSubset` can be used when no equivalence information (e.g., in the form of an `AnalysisState`) is available.

srcarroll commented on mlir/lib/Dialect/Tensor/Transforms/BufferizableOpInterfaceImpl.cpp:
```diff
@@ -1147,13 +1023,117 @@ struct SplatOpInterface
+}
+} // namespace
+
+struct InsertSliceOpSubsetOpInterface
```
I see this is used to attach an interface to `InsertSliceOp`, which I admit I'm unfamiliar with attaching interfaces.  So this is probably a dumb question, but why not just add the `SubsetOpInterface` trait to the `InsertSliceOp` instead?

matthias-springer replied:
This is so that the `MLIRTensorDialect` build target does not have to depend on `MLIRBufferizationDialect`. That's also why the `BufferizableOpInterface` implementations in this file are implemented as external models. It also keeps `TensorOps.cpp` and `TensorOps.td` a bit simpler.

srcarroll reviewed (commented):
most of my questions/comments are about `isEquivalentSubset`.  Obviously I'm having a hard time understanding the structure for this, so I'm hoping to get some clarity on it

matthias-springer committed:
[mlir][bufferization] Empty tensor elimination for materialize_in_destination

This revision adds support for empty tensor elimination to "bufferization.materialize_in_destination" by implementing the `InferDestinationOpInterface`.

Furthermore, the One-Shot Bufferize conflict detection is improved for "bufferization.materialize_in_destination".
